src/1-raw-data/mailinglists/mailman_data_test.go

```diff
@@ -0,0 +1,88 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+	"time"
+)
+
+func TestSetDates(t *testing.T) {
+	// test passing in empty start, empty date, same date, start older than end, not a string
+	today := time.Now().Format("2006-01-02")
+	yesterday := time.Now().AddDate(0, 0, -1).Format("2006-01-02")
+	olderStartDateErrorExample := fmt.Errorf("Start date %v was past end date %v. Update input with different start date.", today, "1915-09-18")
+	tests := []struct {
+		comparisonType string
+		start          string
+		end            string
+		wantStart      string
+		wantEnd        string
+		err            error
+	}{
+		{"Dates empty", "", "", yesterday, today, nil},
+		{"Start date empty", "", "1915-09-18", yesterday, "1915-09-18", olderStartDateErrorExample},
+		{"End date empty", "1865-06-17", "", "1865-06-17", today, nil},
+		{"Start and end dates provided and correct", "1865-06-17", "1915-09-17", "1865-06-17", "1915-09-17", nil},
+	}
+	for _, test := range tests {
+		if gotStart, gotEnd, err := setDates(test.start, test.end); err == test.err {
```

Comment:
There's a helpful construct in testing.T for running subtests:
Wrap the body of the for loop in:

```
t.Run(test.comparisionType, func(t *testing.T) {
....
})
```

This gives the go test runner an indication that there are sub-tests here and will make your output look nice

Reply:
https://godoc.org/testing#T.Run